### DISCUSSION TITLE
External domains resolve with cluster domain causing name resolution failures.

### DISCUSSION THREAD

**Discussion by sirmeepington** [coredns/coredns#5723] [Q&A] [answered]
I've installed a fresh bare-metal cluster using Kubeadm with two Debian 11 nodes, they're both using CoreDNS with flannel as the CNI.

When I publish a pod, it can access other pods and services inside the cluster fine, but when it goes to resolve an external domain it attempts to lookup the domain internally, with the namespace first and then without the namespace, and on the third time it will sometimes resolve properly, following is the logs from CoreDNS  
```
[INFO] 172.17.0.1:57756 - 44118 "A IN api.github.com.<namespace>.svc.cluster.local. udp 59 false 512" NXDOMAIN qr,aa,rd 152 0.000145566s
[INFO] 172.17.0.1:3294 - 65173 "A IN api.github.com.svc.cluster.local. udp 50 false 512" NXDOMAIN qr,aa,rd 143 0.000116096s
[INFO] 172.17.0.1:1234 - 48081 "A IN api.github.com. udp 32 false 512" NOERROR qr,rd,ra 62 0.008756678s
```
I'm using the following CoreDNS config map:
```
.:53 {
    log
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes cluster.local in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf
    cache 30
    loop
    reload
    loadbalance
}
```
The `/etc/resolv.conf` on both nodes have correct external nameservers defined (1.1.1.1 & 1.0.0.1), I can access the domains on the nodes themselves outside of the cluster (via netcat) but inside the cluster they fail to resolve.

**Comment by chrisohaver** [comment]
> when it goes to resolve an external domain it attempts to lookup the domain internally, with the namespace first and then without the namespace…

This is normal k8s behavior, per pods with the “clusterFirst” dns policy.

**Reply by sirmeepington** [reply]
I'm assuming the "Temporary failure in name resolution" in affected pods isn't normal behaviour though..?

**Reply by chrisohaver** [reply]
Not sure what you mean by _temporary_ failure. Can you show what logs look like for one of those failures?  What you have shown so far is normal/expected.

**Reply by sirmeepington** [reply]
Running an application that needs to connect to an external service such as a Discord bot comes back with 
```
Exception in thread "main" net.dv8tion.jda.api.exceptions.ErrorResponseException: -1: java.net.UnknownHostException
        at net.dv8tion.jda.internal.requests.RestActionImpl.complete(RestActionImpl.java:227)
        at net.dv8tion.jda.api.requests.RestAction.complete(RestAction.java:633)
        at net.dv8tion.jda.api.utils.SessionControllerAdapter.getGateway(SessionControllerAdapter.java:86)
        at net.dv8tion.jda.internal.JDAImpl.getGateway(JDAImpl.java:318)
        at net.dv8tion.jda.internal.JDAImpl.login(JDAImpl.java:276)
        at net.dv8tion.jda.internal.JDAImpl.login(JDAImpl.java:268)
        at net.dv8tion.jda.api.JDABuilder.build(JDABuilder.java:1874)
        at com.jagrosh.jmusicbot.JMusicBot.startBot(JMusicBot.java:187)
        at com.jagrosh.jmusicbot.JMusicBot.main(JMusicBot.java:65)
Caused by: java.net.UnknownHostException: discord.com: Temporary failure in name resolution
        at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method)
        at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:929)
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1324)
        at java.net.InetAddress.getAllByName0(InetAddress.java:1277)
        at java.net.InetAddress.getAllByName(InetAddress.java:1193)
        at java.net.InetAddress.getAllByName(InetAddress.java:1127)
        at okhttp3.Dns.lambda$static$0(Dns.java:39)
        at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:185)
        at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:149)
        at okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:84)
        at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:214)
        at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
```

**Reply by chrisohaver** [reply]
Can you show what this looks like in the CoreDNS logs?

**Reply by sirmeepington** [reply]
I've just noticed in the coredns logs, this is happening at times: 
`[INFO] 172.17.0.1:1181 - 55762 "A IN <service>.<namespace>.svc.cluster.local.svc.cluster.local. udp 83 false 512" NXDOMAIN qr,aa,rd 176 0.000279202s` where it duplicates the domain suffix.

**Reply by sirmeepington** [reply]
For the above stack trace, this is what's in the logs:
```
[INFO] 172.17.0.1:12018 - 45437 "A IN discord.com.<namespace>.svc.cluster.local. udp 56 false 512" NXDOMAIN qr,aa,rd 149 0.000132097s
[INFO] 172.17.0.1:6169 - 58802 "A IN discord.com.svc.cluster.local. udp 47 false 512" NXDOMAIN qr,aa,rd 140 0.000083s
```
The third entry of `[INFO] 172.17.0.1:64237 - 48633 "A IN discord.com. udp 29 false 512" NOERROR qr,rd,ra 164 0.007097245s` appears shortly after, but is usually enough to cause the issue. 
Could it be connectivity between the pods and the CoreDNS pods being slow?

**Reply by chrisohaver** [reply]
> where it duplicates the domain suffix.

This is also normal Kubernetes ClusterFirst dns policy behavior.  Not ideal, but its how ClusterFirst pods behave when looking up a name that doesn't have a closing dot on the end to make the name a proper FQDN.

**Reply by chrisohaver** [reply]
> Could it be connectivity between the pods and the CoreDNS pods being slow?

I don't think so - I think if there was a timeout on the 3rd query for example, the client would retry (and we don't see retries). It feels more like buggy client behavior.  Are you using alpine pods by any chance?

**Reply by sirmeepington** [reply]
I think the base used in this Dockerfile is openjdk-8-alpine...

Is this an issue with alpine-based pods specifically?

**Reply by sirmeepington** [reply]
Swapped the base image to `adoptopenjdk/openkjdk8:debian-slim` and the name resolution issues still occur inside the new pod.
On the node, the `/etc/resolv.conf` contains `1.1.1.1` and `1.0.0.1`, so even the forwarding should resolve.

**Reply by sirmeepington** [reply]
Little bit of info that may be useful, the two nodes in the cluster are connected behind an OpenVPN server on the control plane node. I don't think this would affect the external DNS as both nodes have basically the same `/etc/resolv.conf` file contents.

**Comment by chrisohaver** [comment]
Perhaps it is a java dns client peculiarity. Does the issue exist when testing from a pod with dns tools such as dig
e.g. `dig discord.com +search +showsearch`

**Reply by sirmeepington** [reply]
Using the `dnsutils` pod:
```
> kubectl exec -it dnsutils -- dig discord.com +search +showsearch

; <<>> DiG 9.9.5-9+deb8u19-Debian <<>> discord.com +search +showsearch
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 34204
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;discord.com.default.svc.cluster.local. IN A

;; AUTHORITY SECTION:
cluster.local.          30      IN      SOA     ns.dns.cluster.local. hostmaster.cluster.local. 1666949429 7200 1800 86400 30

;; Query time: 0 msec
;; SERVER: 10.96.0.10#53(10.96.0.10)
;; WHEN: Fri Oct 28 14:26:43 UTC 2022
;; MSG SIZE  rcvd: 159


; <<>> DiG 9.9.5-9+deb8u19-Debian <<>> discord.com +search +showsearch
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 59008
;; flags: qr aa rd; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;discord.com.svc.cluster.local. IN      A

;; AUTHORITY SECTION:
cluster.local.          30      IN      SOA     ns.dns.cluster.local. hostmaster.cluster.local. 1666949429 7200 1800 86400 30

;; Query time: 0 msec
;; SERVER: 10.96.0.10#53(10.96.0.10)
;; WHEN: Fri Oct 28 14:26:43 UTC 2022
;; MSG SIZE  rcvd: 151


; <<>> DiG 9.9.5-9+deb8u19-Debian <<>> discord.com +search +showsearch
;; global options: +cmd
;; connection timed out; no servers could be reached
command terminated with exit code 9
```

**Comment by chrisohaver** [comment, marked as answer]
Do you have [nodelocal](https://kubernetes.io/docs/tasks/administer-cluster/nodelocaldns/) dns deployed?

**Reply by sirmeepington** [reply]
No, the only non-default things I have deployed on this cluster are:
- flannel
- metallb
- rabbitmq cluster operator
- ingress-nginx
- the pod that's having name resolution issues when getting an external host

**Reply by chrisohaver** [reply]
OK.  Since CoreDNS logs that it sends a response, that must mean that either the response is getting dropped by something in-between before it reaches the client, or that the client gives up waiting for a response.

> The third entry of `[INFO] 172.17.0.1:64237 - 48633 "A IN discord.com. udp 29 false 512" NOERROR qr,rd,ra 164 0.007097245s` appears shortly after

What is the typical amount of elapsed time between the final `discord.com` response that CoreDNS logs in and the prior search path query?  The response time logged suggests it's responding very quickly, but that doesn't include time for the request to reach coredns, and time for the response to reach the client.

**Reply by sirmeepington** [reply]
Following the CoreDNS logs it seems to be around 7s between these two log lines:

```
[INFO] 172.17.0.1:50723 - 46937 "A IN discord.com.<namespace>.svc.cluster.local. udp 56 false 512" NXDOMAIN qr,aa,rd 149 0.00013758s
[INFO] 172.17.0.1:3908 - 45229 "A IN discord.com. udp 29 false 512" NOERROR qr,rd,ra 164 0.007088377s
```

**Reply by chrisohaver** [reply]
OK thats a very long time, enough for timeouts to occur. 
FYI, you can add timestamps to the logs with `--timestamps` option.

Does this problem only occur with that final query for the bare domain without search domains appended? That's really weird.

Does this happen if you use TCP instead of the default UDP?  
e.g. `dig discord.com +search +showsearch +tcp`

**Reply by sirmeepington** [reply]
`The response time logged suggests it's responding very quickly, but that doesn't include time for the request to reach coredns, and time for the response to reach the client.`

After seeing this I looked at the OpenVPN which links both nodes together and made some changes, notably using `udp` over `tcp` and lowering some other settings that can cause a performance hit. 

It seems to have sped up the responses from CoreDNS to within a second overall.

Thanks a ton for the help.

**Comment by sirmeepington** [comment]
The issue was the OpenVPN client/server connecting the two nodes together had some issues that caused messages between CoreDNS and other pods to take a lot of time. Altering some settings seems to have sped this up quite a bit and it's not throwing any issues at the moment. Hoping it stays like this 🤞
